### DISCUSSION TITLE
Coolprop on Java - Java Wrapper

### DISCUSSION THREAD

**Discussion by myrtp** [CoolProp/CoolProp#2258] [Q&A]
Hello I am new here! I want to use the coolprop library in java for a project. So, I followed the instructions on the official coolprop website, under the java wrapper index (http://www.coolprop.org/dev/coolprop/wrappers/Java/index.html). However, I have encountered some issues.
I downloaded the pre-compiled binaries as referred in the instructions. But afterwards I could not locate the platform-independent.7z , in order to download it. Furthermore, when I tried to download the special Java shared library for my system architecture from the link provided at sourgeforge.net an error message popped up that the file is not available or could not be found. 
Did anyone encounter the same issues and could provide a solution as to how to install the java wrapper? Any help and recommendations would be really appreciated.

**Comment by ibell** [comment]
Related to #2259 , it is very difficult to keep the binaries up-to-date, even with auto-generation in a cloud system. For that reason we recommend you build your own. We will see about making limited sets of them more available, but CoolProp is entirely supported by busy volunteers. 

@jowr - could we add Java and Javascript to the actions? There is a docker container for Javascript at least, and it was integrated before in the buildbot builds and could be pulled into actions I think

**Reply by pkirkham** [reply]
This has been sitting unaddressed for a while -- original post was three years ago!

I had compiled cross-platform Windows, Linux and Mac in the past as per Ian's advice. The hardest part was getting a non-technical Mac colleague to do the Mac compile for me! The downside is that you needed three machines (or Windows + WSL for one machine) to compile everything, and the tooling didn't always work first time -- mostly related to Python and Java paths on different systems etc.

After the 8.0.0 release I decided to upgrade the libraries I use in my Java application and I forked the CoolProp codebase to do this. With a deep breath I commenced compiling on Windows and WSL. Got Windows to work, WSL was failing, and MacOS was a distant objective for when I finally buy myself a Mac Mini. Then I started receiving emails that builds had failed on GitHub. Eh? I don't recall setting anything up. What's going on? Upon investigating I discovered that GitHub now has workflows that can do cross-platform compilation (yes, I'm probably years behind the curve here -- I'm only a hobbyist programmer, not a professional) and these had been brought over in the fork. Huzzah! This was something to look into.

After adapting the CSharp cross-platform workflow I was able to create one for Java. The workflow compiles the native libraries for Windows, Linux and Mac, and generates all the JNI wrappers for them. This seems to work, or at least it did for me in my fork of CoolProp.

I've submitted a pull request #3245 to see if this can't get incorporated into the main branch.

**Reply by ibell** [reply]
Thank you for taking this on! I have a comment on the PR, and it would be nice to bring back the testing to make sure the Java wrappers actually work, and update docs if needed - I looked at the test failure and it looks fixable with some fiddling
